src/protocols/secure_channel/CASESession.cpp

```diff
@@ -258,8 +268,15 @@ class CASESession::WorkHelper
         assertChipStackLockedByCurrentThread();
 
         auto * helper = reinterpret_cast<WorkHelper *>(arg);
-        // Hold strong ptr while work is handled
+        // Hold strong ptr while work is handled, and ensure that helper->mStrongPtr does not keep
+        // holding a reference.
         auto strongPtr(std::move(helper->mStrongPtr));
+        if (!strongPtr)
+        {
+            // This can happen if scheduling AfterWorkHandler failed.  Just grab a strong ref to
+            // handler directly, to fulfill our API contract.
+            strongPtr = helper->mWeakPtr.lock();
```

Comment:
Does this correctly handle concurrency access on the helper ? This is very terse, so it's hard to tell what can be shared-memory accessed across thread.

Reply:
Briefly: I believe that's handled correctly, yes, with one caveat.

If you look a few lines higher, this function has assertChipStackLockedByCurrentThread().  So we know we are on the Matter thread.

We could have gotten here in one of two ways.  Either our background code successfully scheduled `AfterWorkHandler` to run on the Matter thread, and then we are not being touched by the background code at this point anymore.  Or the background code failed to schedule `AfterWorkHandler` and it's being called by our watchdog.  In this case the background code will also not touch this object at this point, except for the `helper->mStrongPtr.reset();` bit in `WorkHandler`.  That should move to before we write to mScheduleAfterWorkFailed.  Not new with this PR, but yes, should be fixed.